### PR TITLE
Adds Http4s server interpreter 

### PR DESCRIPTION
**Pull request by hejfelix**
This is just a first stab -- nothing is optimized and the code is not perfect.

It also bumps Http4s version to 0.20.0-M4 with minor changes to ClientTests. 

Looking forward to review comments! (note, it's Christmas so I'll not be responding quickly)

### CODE REVIEW

**Comment by hejfelix** [comment]
Solves https://github.com/softwaremill/tapir/issues/15

File: build.sbt [hunk 1]
```diff
@@ -118,11 +169,17 @@ lazy val akkaHttpServer: Project = (project in file("server/akka-http-server"))
...
   .settings(commonSettings: _*)
   .settings(
     name := "http4s-server",
-    libraryDependencies ++= Seq("org.http4s" %% "http4s-blaze-server" % http4sVersion)
+    libraryDependencies ++= Seq(
+      "org.http4s" %% "http4s-blaze-server" % http4sVersion,
+      "org.http4s" %% "http4s-dsl"          % http4sVersion
+    ),
+    scalacOptions ++= Seq("-Ypartial-unification") ++ scalacLintOptions,
+    addCompilerPlugin("org.spire-math" %% "kind-projector" % "0.9.8"),
```
Comment (adamw) [inline build.sbt]: this should be provided by sbt-softwaremill, or maybe it doesn't work for you?

File: server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.http4s.{Request, Response}
+import tapir.Endpoint
+import tapir.server.tests.ServerTests
+import tapir.typelevel.ParamsAsArgs
+import cats.implicits._
+import scala.concurrent.ExecutionContext
+
+class Http4sServerTests extends ServerTests[IO] {
+
+  implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.Implicits.global // Is this the one we want?
```
Comment (adamw) [inline server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala]: yes, it's usually safe to use, esp in tests :)

File: build.sbt (hunk 1 above)
Comment (hejfelix) [reply build.sbt]: Kind projector? Didn't check, copied this from my separate repo :)

File: server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala (hunk 2 above)
Comment (hejfelix) [reply server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala]: Maybe we should extract this to a trait? We need it  both here and in client tests.

Comment (adamw) [reply server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala]: I don't think there's any module that depends on cats and is shared between client and server tests?

File: build.sbt (hunk 1 above)
Comment (adamw) [reply build.sbt]: ok, I think all of these (options & kind-projector) are provied by the plugin :)

File: server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala (hunk 2 above)
Comment (hejfelix) [reply server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala]: Http4s is used in client tests?

**Comment by adamw** [comment]
Looks good!

There's a couple of rough edges, but the same is true about the akka-http version ;) So nothing that would prohibit merging. Two things that need to be improved in both is encoding handling (utf-8 by default, but that doesn't have to be the case) and what to do if a header fails to parse.

Unfortunately there's a need for a rebase, as I've change `TypeMapper` to `Codec` and the result isn't always `String` - for bodies it might also be a `Array[Byte]`.

By the way, if you would have any comments, simplifications or suggestions on the design, API, internal API, typeclasses used etc. - it would be great to get another pair of eyes :)

File: server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala (hunk 2 above)
Comment (adamw) [reply server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala]: yes, to provide a server to which the clients can talk to.

**Comment by hejfelix** [comment]
I rebased, but I have to make a real fix for the codec stuff once I have the time. At least it compiles again :)

**Comment by hejfelix** [comment]
Right, now it works :)

File: server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala (hunk 2 above)
Comment (hejfelix) [reply server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala]: Just to be sure — you do agree with bumping the version of http4s to 0.20.0-m4 everywhere?

Comment (adamw) [reply server/http4s-server/src/test/scala/tapir/server/http4s/Http4sServerTests.scala]: yes sure :)

**Pull request merged**